src/Microsoft.AspNet.Mvc.Core/SessionStateTempDataProvider.cs

```
@@ -37,7 +65,27 @@ public class SessionStateTempDataProvider : ITempDataProvider
                 using (var memoryStream = new MemoryStream(value))
                 using (var writer = new BsonReader(memoryStream))
                 {
-                    tempDataDictionary = jsonSerializer.Deserialize<Dictionary<string, object>>(writer);
+                    tempDataDictionary = _jsonSerializer.Deserialize<Dictionary<string, object>>(writer);
+                }
+                foreach (var item in tempDataDictionary.ToList())
+                {
+                    var jArrayValue = item.Value as JArray;
+                    if (jArrayValue != null && jArrayValue.Count > 0)
+                    {
+                        Type returnType = null;
+                        _arrayTypeLookup.TryGetValue(jArrayValue[0].Type, out returnType);
+                        if (returnType != null)
```

This should probably throw if the mapping type could not be found. Silently ignoring might suppress issues in our filtering behavior.